### PR TITLE
TINKERPOP-2133 Implement index lookup for Text predicates

### PR DESCRIPTION
**Pull request by skorikov**
Use index lookup for Text predicates (startingWith,endingWith,containing) in Neo4jGraphStep to improve evaluation performance.

See: https://issues.apache.org/jira/browse/TINKERPOP-2133

### CODE REVIEW

File: neo4j-gremlin/src/main/java/org/apache/tinkerpop/gremlin/neo4j/structure/trait/NoMultiNoMetaNeo4jTrait.java [hunk 1]
```diff
@@ -24,20 +24,20 @@
...
 import org.apache.tinkerpop.gremlin.structure.Property;
 import org.apache.tinkerpop.gremlin.structure.T;
 import org.apache.tinkerpop.gremlin.structure.Vertex;
 import org.apache.tinkerpop.gremlin.structure.VertexProperty;
 import org.apache.tinkerpop.gremlin.structure.util.ElementHelper;
 import org.apache.tinkerpop.gremlin.util.iterator.IteratorUtils;
-import org.neo4j.tinkerpop.api.Neo4jDirection;
-import org.neo4j.tinkerpop.api.Neo4jNode;
-import org.neo4j.tinkerpop.api.Neo4jRelationship;
+import org.neo4j.tinkerpop.api.*;
```
Comment (spmallette) [inline neo4j-gremlin/src/main/java/org/apache/tinkerpop/gremlin/neo4j/structure/trait/NoMultiNoMetaNeo4jTrait.java]: perhaps your IDE did this automatically, but could you please avoid the use of the wildcard import (that's not our code style)?

**Comment by spmallette** [comment]
thanks for orchestrating this pull request as i know you likely needed changes from neo4j-contrib. i think it's ok to bump to this new version of neo4j for TinkerPop 3.4.1 but since it's a major dependency change I will[ email the dev list](https://lists.apache.org/thread.html/7e96fee99fe6ce2539819f153d5841a670d0abd07fd12084a8fa566a@%3Cdev.tinkerpop.apache.org%3E) to see if there are any objections to doing so. 

A few issues:

1. There are dependency conflicts being reported during the build by the enforcer plugin. Those will need to be resolved. Note that you might not have noticed them if you built with just `mvn clean install`. We don't automatically run the full tests for neo4j for a number of reasons, so you need to do `mvn clean install -DincludeNeo4j` - in case you're interested you can find a lot of this kind of information in our [developer documentation](http://tinkerpop.apache.org/docs/current/dev/developer/#building-testing).
2.  is there any reason you didn't also modify `MultiMetaNeo4jTrait`? 
3.  are there tests that validate these changes somehow? There are `TextP` tests in the `HastTest` suite of `gremlin-test` but does anything in neo4j need to be configured for the string matching to work?
4. I think this is a big enough change that we need some upgrade documentation. You can add that [here](https://github.com/apache/tinkerpop/blob/master/docs/src/upgrade/release-3.4.x.asciidoc#tinkerpop-341) - that pattern we use should be easy enough to follow to add a new section.

**Comment by skorikov** [comment]
@spmallette Thanks! You were right, I did not run `mvn clean install` with `-DincludeNeo4j`; sorry for that. I will address the issues and update the pull request.

**Comment by robertdale** [comment]
@skorikov Please include any Neo4j upgrade notes, like we did on [TinkerPop 3.3.1](http://tinkerpop.apache.org/docs/3.4.0/upgrade/#_upgrade_neo4j), in `docs/src/upgrade/release-3.4.x.asciidoc`

**Comment by robertdale** [comment]
@skorikov I took a quick look at the change and I'm concerned about the choice of using Neo4j version 3.4.0.  Let alone that the latest point release for that series is 3.4.11 (12/2018), why not just go to the latest and greatest 3.5.1 (12/2018)?

**Comment by spmallette** [comment]
i  think @skorikov  already had to place some effort into changes in neo4j-contrib to get this to even 3.4.11, so i'm inclined to be content with that version especially since it hasn't been upgraded in a while.  I am merely curious though about 3.5.1 since @robertdale mentions it.... @skorikov was there any reason in particular not to use that latest version?

**Comment by skorikov** [comment]
i guess it is somewhat complicated. neo4j-contrib is not very active, currently, as fas as I know there is only one developer contributing to and maintaining it. the latest neo4j version used in [neo4j-tinkerpop-api-impl](https://github.com/neo4j-contrib/neo4j-tinkerpop-api-impl) was updated to 3.4.0 in november 2018; since then not much has happened.

the feature implemented in this pull request would help us greatly in the project we are currently working in, and my intention was to share it with the community. as @spmallette pointed out, implementing it required changing some dependencies too, which is why new versions of [neo4j-tinkerpop-api](https://github.com/neo4j-contrib/neo4j-tinkerpop-api) and [neo4j-tinkerpop-api-impl](https://github.com/neo4j-contrib/neo4j-tinkerpop-api-impl) were released a few days ago.

however, upgrading neo4j to the latest version (3.5.1) is quite another matter. i am afraid i will not be able to do it in the next days. i will fix the build (regarding enforcer problems) and update the pull request though.

**Comment by spmallette** [comment]
thanks for the explanation. i'm still fine with the version your upgrading to. happy to review again when you've addressed the issues i'd noted.

**Comment by spmallette** [comment]
@skorikov is there any chance you're coming back to this PR any time soon? i'm not trying to rush you, of course, but i wanted you to be aware that we are likely doing a release of 3.4.1 in mid-march in case you'd like to see this change as part of that release.

**Comment by spmallette** [comment]
I took a few minutes to try to resolve the enforcer build problems - neo4j seems highly self conflicted in this packaging. Couldn't seem to resolve it. Unless that can get sorted out, I'm not sure that this neo4j version is going to work in the build.

**Comment by spmallette** [comment]
Note that #1149 should fix the pom.xml issues. Once that merges we should see some build problems go away. Since there hasn't been any activity on this PR in a while, I will likely rebase this PR into a TINKERPOP-2133 branch for further evaluation and then close this PR. Any future PRs related to TINKERPOP-2133 should be based on and directed to that branch. I'm happy to handle this another way if it is presented here....thanks.

**Comment by spmallette** [comment]
#1149 did fix the build problems. I've created this branch from master:

https://github.com/apache/tinkerpop/tree/TINKERPOP-2133

and merged this PR there. That leaves:

1. checking into MultiMetaNeo4jTrait and modifying that as needed
1. checking for tests that validate these changes somehow? There are `TextP` tests in the `HastTest` suite of `gremlin-test` but does anything in neo4j need to be configured for the string matching to work?
1. needs upgrade docs.

closing this now and will comment in the JIRA issue as to where things were left.  thanks for getting this issue this far along @skorikov

**Pull request closed without merging**